core/src/Kokkos_Core_fwd.hpp

```diff
@@ -53,7 +53,9 @@
 #include <impl/Kokkos_Error.hpp>
 #include <impl/Kokkos_Utilities.hpp>
 
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_3
```

Comment:
So far, we have used `KOKKOS_ENABLE_DEPRECATED_CODE` to print deprecations warnings, not to remove deprecated code completely (except for `KOKKOS_CONSTEXPR_14 constexpr` in #4063).
I would want to take this opportunity to discuss/clarify what the macro is supposed to do.

Reply:
Deprecation warnings did not work.  See https://github.com/kokkos/kokkos/pull/3217#discussion_r477417704

Reply:
Hmmm... I just tried to reproduce in https://godbolt.org/z/xq86aMajc and the deprecations seem to work fine for `gcc`, `clang`, `icc` and `msvc`. Would you mind reminding me what the specific issue is?

Reply:
https://godbolt.org/z/f3Ed3x37v

Reply:
It seems to work if you specify the C++ version flag correctly, though: https://godbolt.org/z/KP14aaneW

Reply:
generally KOKKOS_ENABLE_DEPRECATED_CODE should be used to remove deprecated code (if it is not defined the code is removed). Hence the name. We probably should have another option for removing warnings if required: KOKKOS_ENABLE_DEPRECATION_WARNINGS.

Reply:
OK. In that case, I'm fine with this pull request. I'm happy to rework the current behavior of the flags in a follow-up pull request.

Reply:
BTW if you mark the forward declaration as deprecated you will get the warnings in your specialization with GCC (regardless of the version).  I missed you had omitted it.

Reply:
...but other compilers need the specialization to be deprecated: https://godbolt.org/z/WPz9TzMPc

Reply:
https://godbolt.org/z/j8E94dv9W